src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -321,32 +272,56 @@ irep_idt generate_java_generic_typet::build_generic_name(
       existing_generic_type.generic_type_arguments().size() >=
         implicitly_generic_original_class.implicit_generic_types().size(),
       "All implicit generic types must be concretised");
-    auto implicit_generic_type_p =
-      implicitly_generic_original_class.implicit_generic_types().begin();
-    const auto &implicit_generic_types_end =
-      implicitly_generic_original_class.implicit_generic_types().end();
+
+    std::string previous_outer_class_name;
     std::string current_outer_class_name;
+    generic_arguments_name_buildert build_generic_arguments(
+      argument_name_printer);
 
-    while(implicit_generic_type_p != implicit_generic_types_end)
+    // add generic arguments to each generic outer class
+    for(const auto &implicit_generic_type :
+        implicitly_generic_original_class.implicit_generic_types())
     {
+      previous_outer_class_name = current_outer_class_name;
       current_outer_class_name =
-        (*implicit_generic_type_p).get_parameter_class_name();
-      generic_name_buffer << current_outer_class_name;
-      generic_name_buffer << build_generic_name_for_class_arguments(
-        generic_argument_p,
-        implicit_generic_type_p,
-        implicit_generic_types_end,
-        current_outer_class_name);
+        implicit_generic_type.get_parameter_class_name();
+
+      // if the class names do not match, it means that the next generic
+      // outer class is being handled
+      if(current_outer_class_name != previous_outer_class_name)
+      {
+        // add the arguments of the previous generic outer class to the
+        // buffer and reset the builder
+        generic_name_buffer << build_generic_arguments.finalize();
+        generic_arguments_name_buildert build_generic_arguments(
+          argument_name_printer);
+
+        // precondition to check that the implicit generic types are ordered
+        // from the outermost outer class inwards
+        PRECONDITION(
+          has_prefix(current_outer_class_name, previous_outer_class_name));
+
+        // add the remaining part of the current generic outer class name to
+        // the buffer
+        generic_name_buffer << current_outer_class_name.substr(
+          previous_outer_class_name.length());
+      }
+
+      // add an argument to the current generic outer class
+      build_generic_arguments.add_argument(*generic_argument_p);
+      ++generic_argument_p;
     }
+    generic_name_buffer << build_generic_arguments.finalize();
     generic_name_buffer << original_class_name.substr(
-      current_outer_class_name.length(), std::string::npos);
+      current_outer_class_name.length());
   }
   else
   {
     generic_name_buffer << original_class_name;
   }
 
-  // if the original class is generic, add tag for the class itself
+  // if the original class is generic, add generic arguments for the class
+  // itself
   if(is_java_generic_class_type(original_class))
```

Comment:
Suggestion for further improvement: I think you can squash the two loops (implicitly generic and the explicitly generic) together by iterating of the instantiated list and storing an iterator into the implicitly generic types. Then when the implicitly generic iterator reaches the end, you know you are at the real class

Something like:
```C++
current_inner_class inner_class_param_p = implicitly_generic_original_class.implicit_generic_types().begin();
for(const reference_typet &instantiated_param : existing_generic_type.generic_type_arguments())
{
  if(inner_class_param_p == end)
    append the rest of the type
  if(inner_class_param->get_class_name() != old_name)
    finalize the last list
    add the substring part of the name
  add_argument(param);
  if(inner_class_param_p != end)
    ++inner_class_param_p;
}
```

Let me know if that doesn't make sense.

Reply:
Nice! I needed to first create a vector containing all implicit generic and generic types, so that I can iterate over them all, but this still reduced the length of the function by quite a bit and made it more readable I think. Thank you.